### PR TITLE
Only upload 3 sets of test results to codecov (possible workaround for hanging builds)

### LINKED ISSUE #4146: Reduce number of reports sent to codecov
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### Description

Codecov may be less flaky if we send it fewer types of reports.

https://sqlfluff.slack.com/archives/G01G58WRJNR/p1670591318076569?thread_ts=1670548220.440059&cid=G01G58WRJNR

Says @tunetheweb : 
> Suggest we try a Linux, windows and DBT build. That would bring it down from double digits to 3 or 4 and may be enough to make Coverage less flakey.

### Use case

_No response_

### Dialect

n/a

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made
Fixes #4146 

Only require (and upload) coverage reports for:
* Python 3.11
* dbt 1.3.0
* Python Windows

Allegedly, uploading fewer coverage reports makes the coverage portion of the CI build less likely to hang.

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by greg-finley** [comment]
Nit: you have to says "closes #4146" or similar for it to close. But I have just linked it. 🙂

**Comment by greg-finley** [comment]
Mind fixing the few coverage gaps it has now?

File: .github/workflows/ci-tests.yml [hunk 1]
```diff
@@ -41,7 +41,25 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [ '3.7', '3.8', '3.9', '3.10', '3.11' ]
+        python-version: [ '3.7', '3.8', '3.9', '3.10' ]
```
Comment (greg-finley) [inline .github/workflows/ci-tests.yml:44]: Would be nice to put some comment that old versions don't have coverage. So hopefully the next person, when 3.12 comes out, will know to put 3.11 in this list and put 3.12 on line 62.

File: .github/workflows/ci-tests.yml [hunk 2]
```diff
@@ -120,7 +138,42 @@ jobs:
           - 5432:5432
     strategy:
       matrix:
-        dbt-version: [ 'dbt020', 'dbt021', 'dbt100', 'dbt130' ]
+        dbt-version: [ 'dbt020', 'dbt021', 'dbt100' ]
```
Comment (greg-finley) [inline .github/workflows/ci-tests.yml:141]: ditto about comment to help us when bumping this version in the future

**Comment by greg-finley** [comment]
There might be a syntax like `if ...` a certain matrix value, then run coverage. Like how [this page](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#example-expanding-configurations) mentions `- if: ${{ matrix.npm }}`

**Comment by barrywhart** [comment]
The `if` syntax doesn't seem well documented to me. Let's save that for a future PR.

**Comment by barrywhart** [comment]
@greg-finley: Ready for another review. Addressed everything except adding `if`. I'm afraid trying to do that could open a can of worms -- this is all the time I can spare to work on this today.

**Review by greg-finley — APPROVED** [review]
Awesome, looks great!

**Pull request merged**